packages/fuels-contract/src/script.rs

```diff
@@ -280,6 +280,10 @@ impl Script {
 
     /// Execute the transaction in a state-modifying manner.
     pub async fn call(self, fuel_client: &FuelClient) -> Result<Vec<Receipt>, Error> {
+        let chan_info = fuel_client.chain_info().await?;
```

Comment:
Making an additional network trip by calling `.chain_info()`) at every contract call is very wasteful; We're doubling the number of trips per call! 

It's very unlikely that the `chain_info` will change in a short period of time. For instance, when calling `do_something()` `n` times in a row, with some processing time in between them.

I'm not suggesting caching at this layer though -- that'd be overkill. But here's a simple potential solution that's easy to implement: 

1. Extend `Script.call(self, fuel_client: &FuelClient)` to take in the chain_info => `Script.call(self, fuel_client: &FuelClient, chain_info: ChainInfo)`;
2. At the `ContractCallHandler` level, extend it (the struct itself) to include a `pub chain_info: ChainInfo`
3. In `Contract::method_hash()` , call `.chain_info()` to get the chain info just at the time of contract handler creation, then pass it to the contract handler creation part (` Ok(ContractCallHandler {...`) because now `ContractCallHandler` holds a `ChainInfo`;
4. At the call level, no more network trips necessary will happen. `chain_info` checking will happen only at contract call handler creation, which isn't very often, but often enough to possibly catch any change in the chain config.

Reply:
Would it be helpful if we made a separate API query just for fetching the consensus params? Since chain info includes a lot of other transient details such as latest block and peer count.

Reply:
It would, but I wouldn't make it a top priority or anything like that. Definitely a nice to have though! 

My suggestion above is just to avoid extra network latency/trips when you're making sequential calls to the same methods 😅

Reply:
If I implement as you stated, then I have to set the ``method_hash`` function to be ``async``. This further means that this will also have to be ``async``: 

``
     Ok (quote! {
         # doc
         pub fn #name(&self #input) -> #result {
             Contract::method_hash(&self.wallet.get_provider().expect("Provider not set up"), self.contract_id, &self.wallet,
                 #tokenized_signature, #output_param, #arg).expect("method not found (this should never happen)")
         }
     })
``

If I turn this function into an ``async`` function, then every contract function will become ``async``, which requires that ``await`` be called on each one.

I implemented this :

`` let consensus_parameters: ConsensusParameters = 
   wallet.get_provider()?.client.chain_info().await?.consensus_parameters.into();

   let instance = MyContract::new(id.to_string(), wallet.clone(), consensus_parameters);
``

If you think this is ok, I will push branch.

Reply:
Turning `method_hash` into async would be an inconvenience, yes. But I believe you can use `Tokio`'s `block_on` around `chain_info().await` to block it, removing the need to turn the caller function into async. See examples in https://docs.rs/tokio/latest/tokio/runtime/struct.Handle.html#method.block_on

Reply:
:partying_face: I will read this and implement it.